### PR TITLE
Use GetTabletsByCell in healthcheck

### LINKED ISSUE #14277: Feature Request: Reduce number of topo calls from vtgate healthcheck
### Feature Description

In production environments, we observed that vtgates impose a significant amount of load on the topo server, because they read each tablet record every `tablet_refresh_interval` duration, which defaults to 1 minute. It is desirable to lower this to a smaller value (say 30 seconds), to discover replica tablet changes in a more timely fashion, but that increases load on topo even more.
There is one option that is currently available to tune this, and that is the `tablet_refresh_known_tablets` flag. However, this has a downside. It is only usable in static environments where a tablet's host never changes after first deployment. Clearly, this is not usable in a k8s environment where tablets might be rescheduled to a new node at any time for any reason.

There is a way to improve this behavior and reduce the load on the topo. And that is to fetch all the tablet records for a cell in one go, instead of one tablet at a time. This is what the `GetTablets` vtctld RPC already does.
However, there is a tradeoff here. Each supported topo server (etcd and zookeeper right now), has a limit on the request/response size. In fact zookeeper does not actually support the `List` function used by `GetTablets` and we fall back to the one tablet at a time method. If there are a sufficient number of tablets to hit this limit, the topo call will error out, and we would have wasted it. This tradeoff is probably worth it, because of the stark reduction in the number of network calls, and the amount of load on the topo.
In the best case, the number of topo calls becomes O(1) from O(n) where n is the total number of tablets being watched by vtgate. In the worst case, we add 1 additional topo call to what is already O(n).

### Use Case(s)

Vitess clusters with a large number of tablets.

### PR DESCRIPTION
**Pull request by deepthi**
## Description
VTGate's healthcheck module currently calls GetTablet for each tablet alias that it discovers in a cell. Instead we can use GetTabletsForCell to fetch all tablets for a cell at once.

This PR does a few more things:
- GetTabletsForCell now handles the case where the response size violates gRPC limits by falling back to one tablet at a time in case of error.
- Previously, the one tablet at a time method had unlimited concurrency. In this PR we introduce a configuration option for concurrency.
- We pass `topoReadConcurrency` from healthcheck into GetTabletsForCell.
- The behavior of `--refresh_known_tablets` flag is different now. Previously we would not read those tablets at all, now we do read them, but ignore any changes if they are already known.

The basic fix has already been tried in production and shown to reduce the number of `Get` calls from vtgate -> topo from O(n) to O(1).

We can consider deprecating and deleting `--refresh_known_tablets` in a future release. The concerns that originally motivated adding that flag in #3965 are alleviated by fetching all tablets in one call to the topo.

## Related Issue(s)
Fixes #14277 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/topo/tablet.go [hunk 1]
```diff
@@ -295,8 +305,10 @@ func (ts *Server) GetTabletsByCell(ctx context.Context, cellAlias string) ([]*Ta
 	if err != nil || len(listResults) == 0 {
 		// Currently the ZooKeeper and Memory topo implementations do not support scans
 		// so we fall back to the more costly method of fetching the tablets one by one.
-		if IsErrType(err, NoImplementation) {
-			return ts.GetTabletsIndividuallyByCell(ctx, cellAlias)
+		// In the etcd case, it is possible that the response is too large. We also fall
```
Comment (mattlord) [inline go/vt/topo/tablet.go:308]: IIRC, ZK has an even smaller limit so I'd avoid mentioning etcd specifically or make it clear that it's an example. If we do mention etcd specifically, it might also be good to add a link or example, e.g.:
```
rpc error: code = ResourceExhausted desc = grpc: trying to send message larger than max (4195197702 vs. 2147483647)
```

File: go/vt/topo/tablet.go [hunk 2]
```diff
@@ -320,15 +332,15 @@ func (ts *Server) GetTabletsByCell(ctx context.Context, cellAlias string) ([]*Ta
 // directly support the topoConn.List() functionality.
 // It returns ErrNoNode if the cell doesn't exist.
 // It returns (nil, nil) if the cell exists, but there are no tablets in it.
-func (ts *Server) GetTabletsIndividuallyByCell(ctx context.Context, cell string) ([]*TabletInfo, error) {
+func (ts *Server) GetTabletsIndividuallyByCell(ctx context.Context, cell string, opt *GetTabletsByCellOptions) ([]*TabletInfo, error) {
```
Comment (mattlord) [inline go/vt/topo/tablet.go:335]: Same note here about passing the options by reference.

**Review by mattlord — APPROVED** [review]
Nice work on this! Makes for a great improvement.

**Review by vmg — APPROVED** [review]

**Comment by deepthi** [comment]
I'm working on a test case, so we'll merge after that is done.

File: go/vt/topo/tablet.go [hunk 3]
```diff
@@ -503,41 +513,52 @@ func DeleteTabletReplicationData(ctx context.Context, ts *Server, tablet *topoda
...
+// If error is ErrPartialResult, the results in the map are
 // incomplete, meaning some tablets couldn't be read.
 // The map is indexed by topoproto.TabletAliasString(tablet alias).
-func (ts *Server) GetTabletMap(ctx context.Context, tabletAliases []*topodatapb.TabletAlias) (map[string]*TabletInfo, error) {
+func (ts *Server) GetTabletMap(ctx context.Context, tabletAliases []*topodatapb.TabletAlias, opt *GetTabletsByCellOptions) (map[string]*TabletInfo, error) {
 	span, ctx := trace.NewSpan(ctx, "topo.GetTabletMap")
 	span.Annotate("num_tablets", len(tabletAliases))
 	defer span.Finish()
 
-	wg := sync.WaitGroup{}
```
Comment (deepthi) [inline go/vt/topo/tablet.go:515]: In a previous commit I had changed this to use an ErrGroup instead to take advantage of its builtin concurrent limit. However, that changes the behavior of this function, so I have reverted that, and instead opted to implement the concurrency limit using a channel.

File: go/vt/topo/tablet_test.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		{
+			name:    "sharded",
+			tablets: 32,
+			opt:     &topo.GetTabletsByCellOptions{Concurrency: 8},
+		},
+		{
+			name:      "sharded with list error",
+			tablets:   32,
+			opt:       &topo.GetTabletsByCellOptions{Concurrency: 8},
+			listError: topo.NewError(topo.ResourceExhausted, ""),
```
Comment (deepthi) [inline go/vt/topo/tablet_test.go:58]: I added a temporary log line (which I removed before commit) and verified that we in fact call `GetTablet` on every tablet in the presence of this error.

File: go/vt/topo/tablet.go [hunk 5]
```diff
@@ -504,40 +516,51 @@ func DeleteTabletReplicationData(ctx context.Context, ts *Server, tablet *topoda
...
 				log.Warningf("%v: %v", tabletAlias, err)
 				// There can be data races removing nodes - ignore them for now.
 				if !IsErrType(err, NoNode) {
-					someError = NewError(PartialResult, "")
+					return NewError(PartialResult, "")
 				}
 			} else {
+				mu.Lock()
 				tabletMap[topoproto.TabletAliasString(tabletAlias)] = tabletInfo
+				mu.Unlock()
```
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/topo/tablet.go]: This code has changed in a later commit, so resolving this.

File: go/vt/topo/tablet_test.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	"vitess.io/vitess/go/vt/topo/memorytopo"
+)
+
+func TestServerGetTabletsByCell(t *testing.T) {
+	tests := []struct {
+		name    string
+		tablets int
+		opt     *topo.GetTabletsByCellOptions
+	}{
+		{
```
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/topo/tablet_test.go:37]: unit test has been added.

File: go/vt/discovery/topology_watcher_test.go [hunk 7]
```diff
@@ -248,94 +222,77 @@ func checkWatcher(t *testing.T, refreshKnownTablets bool) {
...
 			t.Hostname = tablet.Hostname
 			t.PortMap = tablet.PortMap
 			tablet2 = t
 			return nil
-		}); err != nil {
-			t.Fatalf("UpdateTabletFields failed: %v", err)
-		}
-		if _, err := ts.UpdateTabletFields(context.Background(), tablet.Alias, func(t *topodatapb.Tablet) error {
+		})
+		require.Nil(t, err, "UpdateTabletFields failed")
```
Comment (mattlord) [inline go/vt/discovery/topology_watcher_test.go:231]: Any reason to use require.Nil instead of require.NoError here and in a few other places? It's fine, just curious.

File: go/vt/discovery/topology_watcher_test.go [hunk 8]
```diff
@@ -248,94 +222,77 @@ func checkWatcher(t *testing.T, refreshKnownTablets bool) {
...
-	}
-	if _, err := topo.FixShardReplication(context.Background(), ts, logger, "aa", "keyspace", "shard"); err != nil {
-		t.Fatalf("FixShardReplication failed: %v", err)
-	}
+	require.NoError(t, ts.DeleteTablet(context.Background(), tablet2.Alias))
+	_, err = topo.FixShardReplication(context.Background(), ts, logger, "aa", "keyspace", "shard")
+	require.Nil(t, err, "FixShardReplication failed")
 	tw.loadTablets()
-	checkOpCounts(t, counts, map[string]int64{"ListTablets": 1, "GetTablet": 0, "RemoveTablet": 1})
+	checkOpCounts(t, counts, map[string]int64{"ListTablets": 1, "RemoveTablet": 1})
```
Comment (mattlord) [inline go/vt/discovery/topology_watcher_test.go:287]: It looks like this new call is equivalent to the old one. I tend to prefer the previous as it's more explicit about what we expect (w/o looking at how checkOpCounts works), but up to you.

File: go/vt/discovery/topology_watcher_test.go [hunk 9]
```diff
@@ -248,94 +222,77 @@ func checkWatcher(t *testing.T, refreshKnownTablets bool) {
...
-	if _, ok := allTablets[key]; ok || len(allTablets) != 1 {
-		t.Errorf("fhc.GetAllTablets() = %+v; don't want %v", allTablets, key)
-	}
+	assert.NotContains(t, allTablets, key)
+
 	key = TabletToMapKey(tablet2)
-	if _, ok := allTablets[key]; !ok || len(allTablets) != 1 || !proto.Equal(allTablets[key], tablet2) {
-		t.Errorf("fhc.GetAllTablets() = %+v; want %+v", allTablets, tablet2)
-	}
+	assert.Contains(t, allTablets, key)
```
Comment (mattlord) [inline go/vt/discovery/topology_watcher_test.go:279]: Curious why we use require in some places and assert in others? It's fine, I'm just curious. We want the test to fail immediately on the requires (instead of t.fatal), and we only want to log the failure and cause the eventual test failure (but continue on, instead of t.Errorf) when we use assert?

File: go/vt/topo/tablet.go [hunk 10]
```diff
@@ -503,41 +513,52 @@ func DeleteTabletReplicationData(ctx context.Context, ts *Server, tablet *topoda
...
+		// Previously this was always run with unlimited concurrency, so 32 should be fine.
+		concurrency = 32
+	)
 
-	tabletMap := make(map[string]*TabletInfo)
-	var someError error
+	if opt != nil && opt.Concurrency > 0 {
+		concurrency = opt.Concurrency
+	}
+	sem := make(chan int, concurrency)
```
Comment (mattlord) [inline go/vt/topo/tablet.go:537]: Any reason not to use `golang.org/x/sync/semaphore` here? IMO it's better to use stdlib (even in extended like this one) types when we can and we do use it in various other places already.

```
sem := semaphore.NewWeighted(concurrency)
```

Then we can use acquire and release instead of managing the channel and mutex directly: 
```
sem.Acquire(ctx, 1)
...
sem.Release(1)
```

File: go/vt/topo/tablet.go [hunk 11]
```diff
@@ -503,41 +513,52 @@ func DeleteTabletReplicationData(ctx context.Context, ts *Server, tablet *topoda
...
 
 	for _, tabletAlias := range tabletAliases {
 		wg.Add(1)
 		go func(tabletAlias *topodatapb.TabletAlias) {
 			defer wg.Done()
+			sem <- 1
 			tabletInfo, err := ts.GetTablet(ctx, tabletAlias)
-			mutex.Lock()
+			<-sem
+			mu.Lock()
```
Comment (mattlord) [inline go/vt/topo/tablet.go:546]: If you decide not to use a semaphore, then we should use a defer here for the unlock to be safe.

File: go/vt/topo/memorytopo/memorytopo.go [hunk 12]
```diff
@@ -70,6 +70,9 @@ type Factory struct {
 	// err is used for testing purposes to force queries / watches
 	// to return the given error
 	err error
+	// listErr is used for testing purposed to fake errors from
+	// calls to List.
+	listErr error
```
Comment (mattlord) [inline go/vt/topo/memorytopo/memorytopo.go:75]: It looks like err is meant to be used in the same way. Any reason not to use that for list errors as well? Fine to have the additional one if needed or preferred for some reason — but it feels like we might be adding unnecessary code and complexity.

**Review by mattlord — COMMENTED** [review]
I had some minor comments. The most consequential ones -- although still relatively minor -- were related to using a semaphore and deferring mutex unlocking. Let me know what you think. I can come back to this quickly and re-approve at any time.

File: go/vt/topo/memorytopo/memorytopo.go (hunk 12 above)
Comment (deepthi) [reply go/vt/topo/memorytopo/memorytopo.go:75]: We cannot use `err` because it will fail _every_ call to the topo. We want `List` to fail but the subsequent `Get` to pass. `err` was the first thing I looked at, and it doesn't work for this test case.

File: go/vt/topo/tablet_test.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/topo/memorytopo"
+)
+
```
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/topo/tablet_test.go:29]: Great point. I tell people this all the time and forget myself :)

File: go/vt/discovery/topology_watcher_test.go (hunk 9 above)
Comment (deepthi) [reply go/vt/discovery/topology_watcher_test.go:279]: That is exactly right. And it is intentional. I carefully went through all of them to make sure we are doing it right.
If we are unable to `CreateTablet` that's fatal. But if we are checking conditions on a result, there's no reason to not keep going.

File: go/vt/discovery/topology_watcher_test.go (hunk 7 above)
Comment (deepthi) [reply go/vt/discovery/topology_watcher_test.go:231]: Because the func has two return values, not just a single return value of type error. I'd have preferred the `NoError` form, but it doesn't work for this case.

File: go/vt/discovery/topology_watcher_test.go (hunk 8 above)
Comment (deepthi) [reply go/vt/discovery/topology_watcher_test.go:287]: We don't call `GetTablet` at all any more, so this change reflects that fact. I thought about deprecating and eventually deleting the GetTablet metric, but it didn't seem worth it.
Maybe it _is_ worth changing the tests to explicitly expect 0 calls to `GetTablet` as a way of catching regressions in this behavior.

File: go/vt/discovery/topology_watcher.go [hunk 14]
```diff
@@ -90,14 +90,16 @@ type TopologyWatcher struct {
 	firstLoadChan chan struct{}
 }
 
-// NewTopologyWatcher returns a TopologyWatcher that monitors all
-// the tablets that it is configured to watch, and reloads them periodically if needed.
-// As of now there is only one implementation: watch all tablets in a cell.
-func NewTopologyWatcher(ctx context.Context, topoServer *topo.Server, hc HealthCheck, filter TabletFilter, cell string, refreshInterval time.Duration, refreshKnownTablets bool, topoReadConcurrency int, getTablets func(tw *TopologyWatcher) ([]*topo.TabletInfo, error)) *TopologyWatcher {
+// NewCellTabletsWatcher returns a TopologyWatcher that monitors all
+// the tablets in a cell, and reloads them as needed.
+func NewCellTabletsWatcher(ctx context.Context, topoServer *topo.Server, hc HealthCheck, f TabletFilter, cell string, refreshInterval time.Duration, refreshKnownTablets bool, topoReadConcurrency int) *TopologyWatcher {
```
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/discovery/topology_watcher.go:95]: OK. While I'm at it, I'm removing some more indirection as well. If someone comes along who really wants to re-use this, stuff can be factored out again.

File: go/vt/topo/tablet_test.go [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+					Alias: &topodatapb.TabletAlias{
+						Cell: cell,
+						Uid:  uint32(i),
+					},
+					Hostname: "host1",
+					PortMap: map[string]int32{
+						"vt": int32(i),
+					},
+					Keyspace: "keyspace",
+					Shard:    "shard",
```
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/topo/tablet_test.go:95]: Good catch. Fixing it.

File: go/vt/topo/tablet_test.go [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+			listError: topo.NewError(topo.ResourceExhausted, ""),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			const cell = "zone1"
```
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/topo/tablet_test.go:67]: None at all. Moving them out.

File: go/vt/topo/tablet.go (hunk 10 above)
Comment (deepthi) [reply go/vt/topo/tablet.go:537]: Good idea!

File: go/vt/discovery/topology_watcher_test.go (hunk 7 above)
Comment (mattlord) [reply go/vt/discovery/topology_watcher_test.go:231]: I meant that you can use this instead (in the other cases you're just using the `error` return value directly from the function):
```
require.NoError(t, err, "UpdateTabletFields failed")
```

AFAIK they are equivalent, but it's a little more explicit. Not a problem at all either way.

**Comment by deepthi** [comment]
@mattlord I think I've addressed all the comments. This is ready for final review. If it looks good to you, please remove the "Do Not Merge" label.

File: go/vt/topo/tablet.go (hunk 11 above)
Comment (mattlord) [reply go/vt/topo/tablet.go:557]: If we do keep the mutex, we should unlock it in a defer here as we release it at the end of the closure/goroutine anyway.

**Review by mattlord — COMMENTED** [review]
I have some minor notes -- sorry to keep going back and forth on this. IMO it's worth being extra cautious in this case though. Let me know what you think -- maybe I'm being overly so.

File: go/vt/topo/tablet.go [hunk 17]
```diff
@@ -503,41 +515,61 @@ func DeleteTabletReplicationData(ctx context.Context, ts *Server, tablet *topoda
...
+	}
+	var sem = semaphore.NewWeighted(concurrency)
 
 	for _, tabletAlias := range tabletAliases {
 		wg.Add(1)
 		go func(tabletAlias *topodatapb.TabletAlias) {
 			defer wg.Done()
+			if err := sem.Acquire(ctx, 1); err != nil {
+				// Only happens if context is cancelled.
+				mu.Lock()
```
Comment (mattlord) [inline go/vt/topo/tablet.go:547]: I think that we should do this:
```
mu.Lock()
defer mu.Unlock()
...
return
```

We definitely want to return in this block as we didn't acquire the semaphore so we don't want to continue on in the goroutine.

Comment (mattlord) [reply go/vt/topo/tablet.go:547]: Might be getting hard to follow at this point, so this is what I was thinking: https://gist.github.com/mattlord/089986f60bffe33d1d7b3faa21942e06

File: go/vt/topo/tablet.go [hunk 18]
```diff
@@ -282,21 +284,31 @@ func (ts *Server) GetTabletAliasesByCell(ctx context.Context, cell string) ([]*t
 	return result, nil
 }
 
+// GetTabletsByCellOptions controls the behavior of
+// Server.FindAllShardsInKeyspace.
+type GetTabletsByCellOptions struct {
+	// Concurrency controls the maximum number of concurrent calls to GetTablet.
+	// For backwards compatibility, concurrency of 0 is considered unlimited.
```
Comment (mattlord) [inline go/vt/topo/tablet.go:291]: I don't think this is right anymore, is it? I say that because we currently only use the value if it's > 0.

Comment (deepthi) [reply go/vt/topo/tablet.go:291]: Yeah.. this has gone through 3 different implementations at this point, and somewhere in there we lost the infinite concurrency. I'm inclined to think 32 is fine as a default and we don't need to worry about infinite concurrency at all 🤷

File: go/vt/topo/tablet.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/topo/tablet.go:557]: I tried replacing this with FirstErrorRecorder. The code is no less awkward, because we don't actually want to record the returned error, we want to wrap it in a different error. And FirstErrorRecorder wraps its own mutex so you are not actually saving on that.

```
	var errRecorder concurrency.FirstErrorRecorder

	for _, tabletAlias := range tabletAliases {
		wg.Add(1)
		go func(tabletAlias *topodatapb.TabletAlias) {
			defer wg.Done()
			if err := sem.Acquire(ctx, 1); err != nil {
				// Only happens if context is cancelled.
				log.Warningf("%v: %v", tabletAlias, err)
				// We only need to set this on the first error.
				errRecorder.RecordError(NewError(PartialResult, tabletAlias.GetCell()))
				return
			}
			tabletInfo, err := ts.GetTablet(ctx, tabletAlias)
			sem.Release(1)
			if err != nil {
				log.Warningf("%v: %v", tabletAlias, err)
				// There can be data races removing nodes - ignore them for now.
				// We only need to set this on first error.
				if !IsErrType(err, NoNode) {
					errRecorder.RecordError(NewError(PartialResult, tabletAlias.GetCell()))
				}
			} else {
				mu.Lock()
				tabletMap[topoproto.TabletAliasString(tabletAlias)] = tabletInfo
				mu.Unlock()
			}
		}(tabletAlias)
	}
```

`sync.Once` is also problematic. The two places where we want to catch the error are slightly different and don't fit into a neat function. I had the same problem with trying to move the lock + logic + unlock into a closure.

File: go/vt/topo/tablet.go (hunk 17 above)
Comment (deepthi) [reply go/vt/topo/tablet.go:547]: Oh yeah I was looking at the code again and realized I forgot a return 😮‍💨 
Glad you caught it too. I think we actually need a lot more test coverage than we have, but I'm calling that debt at this point. We are slightly better off than before given that at least I added 4 unit tests.
Error conditions are usually the hardest to test, and that is where most the missing coverage tends to be.

File: go/vt/topo/tablet.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/topo/tablet.go:557]: At least I was able to change both calls to Unlock to be deferred calls, so that's a win.

File: go/vt/topo/tablet.go (hunk 17 above)
Comment (mattlord) [reply go/vt/topo/tablet.go:546]: You can use different functions in sync.Once.Do, it will only execute the Do method once. But it doesn't matter. It's not much different than what is there now.
